File: android/src/main/java/com/plugins/forgerockbridge/ForgotPasswordHandler.java
package com.plugins.forgerockbridge;

import android.content.Context;
import android.util.Log;

import com.getcapacitor.JSObject;
import com.getcapacitor.PluginCall;


import org.forgerock.android.auth.FRSession;

import org.forgerock.android.auth.NodeListener;

import org.forgerock.android.auth.callback.PasswordCallback;
import org.forgerock.android.auth.callback.ValidatedPasswordCallback;
import org.forgerock.android.auth.Node;
import org.forgerock.android.auth.callback.Callback;

import com.plugins.forgerockbridge.nodeListenerCallbacks.ForgotPasswordNodeListener;
import com.plugins.forgerockbridge.state.PluginState;

public class ForgotPasswordHandler {
    private static final String TAG = "ForgeRockBridge";

    public static void startForgotPasswordJourney(PluginCall call, Context context, NodeListener<FRSession> listener ) {
        Log.d(TAG, "[ForgotPasswordHandler: startForgotPasswordJourney ] call method startForgotPasswordJourney");


        String journey = call.getString("journey");
        String username = call.getString("username");

        if (journey == null || journey.trim().isEmpty()) {
            Log.e(TAG, "[ForgotPasswordHandler: startForgotPasswordJourney ]: Error in get Journey");
            ErrorHandler.reject(call, ErrorHandler.ErrorCode.MISSING_JOURNEY);
            return;
        }

         if (username == null || username.trim().isEmpty()) {
            Log.e(TAG, "[ForgotPasswordHandler: startForgotPasswordJourney]: Error in get the username");
            ErrorHandler.reject(call, ErrorHandler.ErrorCode.MISSING_JOURNEY);
            return;
        }

        FRSession.authenticate(context, journey, listener);
    }

    public static void getSecurityQuestion(PluginCall call, PluginState state) {
        Node pending = state.getPendingNode();
        String question = null;

        if (pending == null) {
            Log.e(TAG, "[ForgotPasswordHandler: getSecurityQuestion]: NO PENDING NODE SAVE");
            ErrorHandler.reject(call, ErrorHandler.ErrorCode.NO_PENDING_NODE, "NO PENDING NODE SAVE");
            return;
        }


        for (Callback cb : pending.getCallbacks()) {
            if (cb instanceof PasswordCallback) {
                question = ((PasswordCallback) cb).getPrompt();
                break;
            }
        }

        if (question == null || question.trim().isEmpty()) {
            Log.e(TAG, "[ForgotPasswordHandler: getSecurityQuestion]: NO QUESTION FOUND");
            ErrorHandler.reject(call, ErrorHandler.ErrorCode.NO_QUESTION_FOUND, "NO QUESTION FOUND");
            return;
        }

        JSObject out = new JSObject()
            .put("question", question);
            Log.d(TAG, "[ForgotPasswordHandler: getSecurityQuestion] Sending question");
        call.resolve(out);
    }

    public static void answerQuestionForgotPassword(PluginCall call, PluginState state, Context context ) {
        Node pending = state.getPendingNode();
        String answer = call.getString("answer");

        if (pending == null) {
            Log.e(TAG, "[ForgotPasswordHandler: answerQuestionForgotPassword]: NO PENDING NODE SAVE");
            ErrorHandler.reject(call, ErrorHandler.ErrorCode.NO_PENDING_NODE, "NO PENDING NODE SAVE");
            return;
        }

        if (answer == null || answer.trim().isEmpty()) {
            Log.e(TAG, "[ForgotPasswordHandler: answerQuestionForgotPassword]: MISSING ANSWER");
            ErrorHandler.reject(call, ErrorHandler.ErrorCode.MISSING_PARAMETER, "Missing answer");
            return;
        }


        for (Callback cb : pending.getCallbacks()) {
            if (cb instanceof PasswordCallback) {
                ((PasswordCallback) cb).setPassword(answer.toCharArray());
                break;
            } 
        }

        pending.next(context, new ForgotPasswordNodeListener(call, context, state));
    }

    public static void changePasswordForgotPassword (PluginCall call, PluginState state, Context context ) {
        Log.d(TAG, "[ForgotPasswordHandler: changePasswordForgotPassword]: CALL METHOD changePasswordForgotPassword");
        Node pending = state.getPendingNode();
        String password = call.getString("password", null);

        if (pending == null) {
            Log.e(TAG, "[ForgotPasswordHandler: changePasswordForgotPassword]: NO PENDING NODE SAVE");
            ErrorHandler.reject(call, ErrorHandler.ErrorCode.NO_PENDING_NODE, "NO PENDING NODE SAVE");
            return;
        }

        if (password == null || password.trim().isEmpty()) {
            Log.e(TAG, "[ForgotPasswordHandler: changePasswordForgotPassword]: MISSING PASSWORD");
            ErrorHandler.reject(call, ErrorHandler.ErrorCode.MISSING_PARAMETER, "Missing password");
            return;
        }

        for (Callback cb : pending.getCallbacks()) {
            if (cb instanceof ValidatedPasswordCallback) {
                ((ValidatedPasswordCallback) cb).setPassword(password.toCharArray());
                break;
            }
        }

        pending.next(context, new ForgotPasswordNodeListener(call, context, state));
    }

}

File: ios/Sources/ForgerockBridgePlugin/ForgotPasswordHandler.swift
import Foundation
import Capacitor
import FRAuth
import FRCore
import FRAuthenticator

class ForgotPasswordHandler {
    private let call: CAPPluginCall
    private let plugin: ForgerockBridgePlugin

    init(call: CAPPluginCall, plugin: ForgerockBridgePlugin) {
        self.call = call
        self.plugin = plugin
    }

    func startForgotPasswordJourney (_ call: CAPPluginCall,completion: @escaping NodeCompletion<Token> ) {
        print("[ForgotPasswordHandler:startForgotPasswordJourney] CALL startForgotPasswordJourney")
        
        plugin.pendingNode = nil
        
        guard let journey = call.getString("journey") else {
            ErrorHandler.reject(call, code: ErrorCode.missingJourney, message: "MISSING JOURNEY")
            return
        }

        guard let username = call.getString("username") else {
            ErrorHandler.reject(call, code: ErrorCode.missingParameter, message: "MISSING USERNAME")
            return
        }
        
          FRSession.authenticate(authIndexValue: journey, completion: completion)
    }

    func getSecurityQuestion (_ call: CAPPluginCall) {
        print("[ForgotPasswordHandler:getSecurityQuestion] CALL getSecurityQuestion")
        
        var question: String? = nil

        guard let pending = self.plugin.pendingNode else {
            print("[getSecurityQuestion] NO PENDING NODE SAVE")
            ErrorHandler.reject(call, code: .noPendingNode, message: "NO PENDING NODE SAVE")
            return
        }
        
        for (_, callback) in pending.callbacks.enumerated() {
            if let passwordCallback = callback as? PasswordCallback {
                question = passwordCallback.prompt
                break
            }
        }
        
        guard let q = question else {
            print("[getSecurityQuestion] NO QUESTION FOUND")
            ErrorHandler.reject(call, code: .noQuestionFound, message: "NO QUESTION FOUND")
            return
        }
        
        let out: [String: Any] = [
                    "question": q
                ]
                
        print("[getSecurityQuestion] Sending question")
        call.resolve(out)
    }
    
    func answerQuestionForgotPassword (_ call: CAPPluginCall, completion: @escaping NodeCompletion<Token> ) {
        
        guard let pending = self.plugin.pendingNode else {
            print("[answerQuestionForgotPassword] NO PENDING NODE SAVE")
            ErrorHandler.reject(call, code: .noPendingNode, message: "NO PENDING NODE SAVE")
            return
        }
        
        guard let answer = call.getString("answer") else {
            print("[answerQuestionForgotPassword] MISSING PARAMETER answer")
            ErrorHandler.reject(call, code: ErrorCode.missingParameter, message: "MISSING PARAMETER answer")
            return
        }
        
        for (_, callback) in pending.callbacks.enumerated() {
            if let passwordCallback = callback as? PasswordCallback {
                passwordCallback.setValue(answer)
                break
            } else if let confirmationCallback = callback as? ConfirmationCallback {
                confirmationCallback.value = 0
            }
        }
        
        pending.next(completion: completion)
        
    }

    func changePasswordForgotPassword (_ call: CAPPluginCall,completion: @escaping NodeCompletion<Token> ) {

        guard let pending = self.plugin.pendingNode else {
            print("[changePasswordForgotPassword] NO PENDING NODE SAVE")
            ErrorHandler.reject(call, code: .noPendingNode, message: "NO PENDING NODE SAVE")
            return
        }

        guard let password = call.getString("password") else {
            print("[changePasswordForgotPassword] MISSING PARAMETER password")
            ErrorHandler.reject(call, code: ErrorCode.missingParameter, message: "MISSING PARAMETER password")
            return
        }

        for (_, callback) in pending.callbacks.enumerated() {
            if let validated = callback as? ValidatedCreatePasswordCallback {
                print("[changePasswordForgotPassword] CALLBACK ValidatedCreatePasswordCallback Detected")
                validated.setValue(password)
                break
            }
        }
        
        pending.next(completion: completion)

    }
}
